### PR TITLE
BUG: is_scalar_indexer

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

ATM we go through the wrong path when setting a listlike entry in a single position.

The broader goal: separate out the cases where setitem is inplace vs copying.

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
Is there an actual bug that this fixes? (do you have an example with indexing that now did the wrong thing?)

File: pandas/core/indexers.py [hunk 1]
```diff
@@ -65,18 +65,20 @@ def is_list_like_indexer(key) -> bool:
     return is_list_like(key) and not (isinstance(key, tuple) and type(key) is not tuple)
 
 
-def is_scalar_indexer(indexer, arr_value) -> bool:
+def is_scalar_indexer(indexer, ndim: int) -> bool:
```
Comment (simonjayhawkins) [inline pandas/core/indexers.py]: can you add a parameters section to the docstring while in the neighbourhood and making changes to the parameters

**Comment by jbrockmendel** [comment]
> Is there an actual bug that this fixes? (do you have an example with indexing that now did the wrong thing?)

Nothing user-facing, just an unnecessary failed-cast:

```
df = DataFrame(index=[0, 1], columns=[0])
df.iloc[1, 0] = [1, 2, 3]
```

ATM this incorrectly doesnt see the indexer as a scalar-indexer, but does see this (also incorrectly) as an `exact_match`, so goes through the try/except in L906-909.  This is why the try/except is needed in the first place.

File: pandas/core/internals/blocks.py [hunk 2]
```diff
@@ -892,12 +892,10 @@ def setitem(self, indexer, value):
...
         # then use the resultant dtype
         elif exact_match:
+            # We are setting _all_ of the array's values, so can cast to new dtype
             values[indexer] = value
 
-            try:
-                values = values.astype(arr_value.dtype)
-            except ValueError:
-                pass
+            values = values.astype(arr_value.dtype)
```
Comment (jreback) [inline pandas/core/internals/blocks.py]: can add copy=False here I think

**Review by jreback — COMMENTED** [review]
small comment, otherwise lgtm.

**Review by jreback — APPROVED** [review]
lgtm. merge on green.

**Pull request merged**